components/brave_wallet/browser/zcash/rust/orchard_shard_tree_impl.cc

```diff
@@ -0,0 +1,347 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/zcash/rust/orchard_shard_tree_impl.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/memory/ptr_util.h"
+#include "base/ranges/algorithm.h"
+#include "brave/components/brave_wallet/browser/internal/orchard_storage/zcash_orchard_storage.h"
+#include "brave/components/brave_wallet/browser/zcash/rust/cxx_orchard_shard_tree_delegate.h"
+#include "brave/components/brave_wallet/browser/zcash/rust/lib.rs.h"
+#include "brave/components/brave_wallet/browser/zcash/rust/orchard_decoded_blocks_bundle_impl.h"
+#include "brave/components/brave_wallet/common/zcash_utils.h"
+
+namespace brave_wallet::orchard {
+
+::brave_wallet::OrchardShardAddress FromRust(const ShardTreeAddress& addr) {
+  return ::brave_wallet::OrchardShardAddress{addr.level, addr.index};
+}
+
+ShardTreeAddress ToRust(const ::brave_wallet::OrchardShardAddress& addr) {
+  return ShardTreeAddress{addr.level, addr.index};
+}
+
+ShardTreeCap ToRust(::brave_wallet::OrchardShardTreeCap& shard_store_cap) {
+  ::rust::Vec<uint8_t> data;
+  data.reserve(shard_store_cap.size());
+  base::ranges::copy(shard_store_cap, std::back_inserter(data));
+  return ShardTreeCap{std::move(data)};
+}
+
+::brave_wallet::OrchardShardTreeCap FromRust(const ShardTreeCap& cap) {
+  ::brave_wallet::OrchardShardTreeCap shard_store_cap;
+  shard_store_cap.reserve(cap.data.size());
+  base::ranges::copy(cap.data, std::back_inserter(shard_store_cap));
+  return shard_store_cap;
+}
+
+::brave_wallet::OrchardShard FromRust(const ShardTreeShard& tree) {
+  std::optional<OrchardShardRootHash> shard_root_hash;
+  if (!tree.hash.empty()) {
+    CHECK_EQ(kOrchardShardTreeHashSize, tree.hash.size());
+    OrchardShardRootHash hash_value;
+    base::ranges::copy(tree.hash, hash_value.begin());
+    shard_root_hash = hash_value;
+  }
+
+  std::vector<uint8_t> data;
+  data.reserve(tree.data.size());
+  base::ranges::copy(tree.data, std::back_inserter(data));
+
+  return ::brave_wallet::OrchardShard(FromRust(tree.address), shard_root_hash,
+                                      std::move(data));
+}
+
+ShardTreeShard ToRust(const ::brave_wallet::OrchardShard& tree) {
```

Comment:
I think we need to work on these type names because it's very confusing to me to convert between a type called `OrchardShard` and `ShardTreeShard`. We seem to have a bunch of different naming conventions going on here so I think for now let's standardize on using `CXX` prefix for the rust types

Reply:
Renamed structs and Result<Optional> wrappers